irohad/ordering/impl/ordering_service_impl.cpp

```diff
@@ -33,55 +33,99 @@ namespace iroha {
         size_t delay_milliseconds,
         std::shared_ptr<network::OrderingServiceTransport> transport,
         std::shared_ptr<ametsuchi::OrderingServicePersistentState>
-            persistent_state)
+            persistent_state,
+        bool is_async)
         : wsv_(wsv),
           max_size_(max_size),
           delay_milliseconds_(delay_milliseconds),
           transport_(transport),
-          persistent_state_(persistent_state),
-          is_finished(false) {
-      updateTimer();
+          persistent_state_(persistent_state) {
       log_ = logger::log("OrderingServiceImpl");
 
       // restore state of ordering service from persistent storage
-      proposal_height = persistent_state_->loadProposalHeight().value();
+      proposal_height_ = persistent_state_->loadProposalHeight().value();
+
+      rxcpp::observable<ProposalEvent> timer =
+          rxcpp::observable<>::interval(
+              std::chrono::milliseconds(delay_milliseconds_),
+              rxcpp::observe_on_new_thread())
+              .map([](auto) { return ProposalEvent::kTimerEvent; });
+
+      if (is_async) {
+        handle_ =
+            rxcpp::observable<>::from(timer, transactions_.get_observable())
+                .merge(rxcpp::synchronize_new_thread())
+                .subscribe([this](auto &&v) {
+                  auto check_queue = [&] {
+                    switch (v) {
+                      case ProposalEvent::kTimerEvent:
+                        return not queue_.empty();
+                      case ProposalEvent::kTransactionEvent:
+                        return queue_.unsafe_size() >= max_size_;
+                      default:
+                        BOOST_ASSERT_MSG(false, "Unknown value");
```

Comment:
I think better say smth to log & return false
Also does the compilers enforce `default` (e.g like not all cases handled or so)? so maybe just remove it

Reply:
Why? If a wrong value of enum class somehow appeared in observable, then someone has manually casted a value to enum, which should not happen.